defi/single-token-amm-dex/src/lib.rs

```diff
@@ -0,0 +1,904 @@
+                let entry_fee = a;
+                let candy_addr = c;               
+                
+                let output_bucket: Bucket;
+                
+                if flag == 0i32.into() {
+                    m_candy_amnt = m_candy_amnt+meta_amount;
+                    self.badge_map.insert(meta_badge_addr,(entry_fee,m_candy_amnt,candy_addr));
+                    output_bucket = meta_badge;
+                } else if flag == 1i32.into() {
+                    assert!( meta_amount <= m_candy_amnt," Let's check passed amount ");
+                    m_candy_amnt = m_candy_amnt-meta_amount;
+                  
+                    if m_candy_amnt == 0i32.into() {
+                        self.badge_map.remove(&meta_badge_addr);
+                        CandyDex::badge_burn(self, meta_badge);
+                        output_bucket = Bucket::new(RADIX_TOKEN);
+                    } else { 
+                        self.badge_map.insert(meta_badge_addr,(entry_fee,m_candy_amnt,candy_addr));
+                        output_bucket = meta_badge; 
+                    }
+
+                } else { 
+                    std::process::abort(); 
+                }
+                output_bucket
+            }
+            // Create a metaCandy resource relative to a kind of candy provided to protocol by end 
+            // users.
+            fn add_meta_candy(
+                &mut self, 
+                name: String, 
+                symbol: String, 
+                address: Address
+            ) -> Address {
+                assert!(!self.meta.contains_key(&address)," Candy already exist ");
+                
+                let meta_res_def = ResourceBuilder::new_fungible(DIVISIBILITY_MAXIMUM)
+                    .metadata("name", format!(" m{}", name.clone()))
+                    .metadata("symbol", format!(" m{}", symbol.clone()))
+                    .flags(MINTABLE | BURNABLE)
+                    .badge(self.minter_badge.resource_def(), MAY_MINT | MAY_BURN)
+                    .no_initial_supply();
+
+                let meta_address = meta_res_def.address();   
+
+                self.meta.insert(address.clone(),MetaToken::new(name, symbol, address, meta_res_def));
+                
+                let (a,b,_c) = self.meta_map.get(&address.clone()).unwrap();               
+                let accrued_fee = *a;
+                let meta_minted = *b;
+                self.meta_map.insert(address,(accrued_fee,meta_minted,meta_address));
+                
+                meta_address
+            }
+
+            // Mint a metacandy amount relative to an amount of candy provided to protocol and 
+            // update total metaCandy minted amount.
+            fn meta_mint(&mut self, candy_amnt: Decimal, candy_address: Address) -> Bucket {             
+                let temp_badge = self.minter_badge.take(1);                
```

Comment:
There is a shortcut, instead of having to take from the vault, minting, then putting back, do this:

```
let meta_candy = self.minter_badge.authorize(|auth| {
  m_cnady.meta_res_def.min(candy_amnt, auth)
});
```